Expose cdrom in UI

glekner opened the pull request:
https://jira.coreos.com/browse/KNIP-520 - Story 1

rawagner commented on frontend/packages/kubevirt-plugin/src/selectors/selectors.ts:
```diff
@@ -12,6 +13,8 @@ export const getAnnotations = (vm: VMLikeEntityKind, defaultValue?: any) =>
 export const getDescription = (vm: VMLikeEntityKind) =>
   _.get(vm, 'metadata.annotations.description');
 
+export const getCdRom = (vm: VMLikeEntityKind) => getDisks(vm).filter((disk) => !!disk.cdrom);
```
I'd call this `getCDRoms` as you are returning an array.

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -93,6 +94,8 @@ export const VMDetailsList: React.FC<VMResourceListProps> = ({ vm, vmi, pods, mi
         )}
       </dd>
+      <dt>CD</dt>
+      <dd id={prefixedID(id, 'cd-rom')}>{cd.length > 0 ? cd[0].name : 'No Disk'}</dd>
```
we only first one and not all of them ?

rawagner replied:
OK, i see that the design shows only one CD-ROM but we can have more @matthewcarleton

glekner replied:
So there could be multiple Cds? ok, should we display them in some kind of order? @rawagner

rawagner replied:
according to docs https://kubevirt.io/user-guide/docs/latest/creating-virtual-machines/disks-and-volumes.html#cdrom any disk with 'cdrom' attribute is CDRom and it does not mention we can have only one. CDRom can also have boot order assigned (as any disk) so maybe we want to order them by boot order, then just by position in yaml ?

glekner commented:
- Now using `getBootableDevicesInOrder` function to display the CD or multiple CDs sorted by `bootOrder`

glekner commented:
Reverted back to display only one CD.
after discussing with @jelkosz its by design

rawagner commented on frontend/packages/kubevirt-plugin/src/selectors/selectors.ts:
```diff
@@ -12,6 +13,8 @@ export const getAnnotations = (vm: VMLikeEntityKind, defaultValue?: any) =>
 export const getDescription = (vm: VMLikeEntityKind) =>
   _.get(vm, 'metadata.annotations.description');
 
+export const getCdRom = (vm: VMLikeEntityKind) => getDisks(vm).filter((device) => !!device.cdrom);
```
ok, but still, we should call this function `getCDRoms` (plural and `CD` capital letters).

glekner replied:
fixed

glekner commented:
updated to support multiple CDs

glekner commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -93,6 +94,10 @@ export const VMDetailsList: React.FC<VMResourceListProps> = ({ vm, vmi, pods, mi
       </dd>
+      <dt>CD</dt>
+      <dd id={prefixedID(id, 'cd-rom')}>
+        {sortedCds.length > 0 ? <BootOrder bootableDevices={sortedCds} /> : 'No Disk'}
```
@rawagner seems reasonable to use `BootOrder` to display the multiple CDs in boot order?

rawagner replied:
yes, it does!

But we are trying to get rid of kubevirt-web-ui-components dependencies so I'd suggest to port `BootOrder` and `getBootableDevicesInOrder` to `kubevirt-plugin`.

I personally would expect that `BootOrder` will order the devices itself (it will call `getBootableDevicesInOrder` on `bootableDevices` prop)

glekner commented:
#2477 is almost ready for review.

meanwhile, updated this PR to display the CDs info as intended.
can you take a look @matthewcarleton 

--------------
**Windows VM**
<img width="502" alt="Screen Shot 2019-09-15 at 15 08 51" src="https://user-images.githubusercontent.com/24938324/64921365-1d373100-d7cb-11e9-9e9c-8add71b09bc9.png">

------------
**Multiple CDS**
<img width="285" alt="Screen Shot 2019-09-15 at 15 01 30" src="https://user-images.githubusercontent.com/24938324/64921370-20322180-d7cb-11e9-8bbf-3798403d22e4.png">

matthewcarleton commented:
@glekner This looks good! 
I wonder if we can simplify it a bit by not including the entire string for the windows tool. It's provided (and read only) right? So maybe just showing the name of the qcow2 file instead could work. WDYT?

glekner commented:
![image](https://user-images.githubusercontent.com/24938324/64943615-e4ec2d00-d874-11e9-8a8e-6f9ce139fb7a.png)

@matthewcarleton looks better, still takes two lines.

matthewcarleton commented:
@glekner what if we did this:
<img width="238" alt="Screen Shot 2019-09-16 at 9 11 58 AM" src="https://user-images.githubusercontent.com/287880/64956987-3b03a500-d862-11e9-82b3-18eab81cc5c5.png">

glekner commented:
> @glekner what if we did this:
> <img alt="Screen Shot 2019-09-16 at 9 11 58 AM" width="238" src="https://user-images.githubusercontent.com/287880/64956987-3b03a500-d862-11e9-82b3-18eab81cc5c5.png">

I really like it, but wouldn't it cause inconsistency with the other sections?

matthewcarleton commented:
> I really like it, but wouldn't it cause inconsistency with the other sections?

how so? Because of the bold text? I think it's justified b/c CDROMs are the only ones with a title/content format.

glekner commented:
> > I really like it, but wouldn't it cause inconsistency with the other sections?
> 
> how so? Because of the bold text? I think it's justified b/c CDROMs are the only ones with a title/content format.

Sounds good to me, ill implement it.

matthewcarleton commented:
@glekner I wonder if we should drop the numbers for this list too. @andybraren made a good point that this could be a bit confusing b/c we see this in the boot order and it's there to indicate the order. I don't believe we need to indicate order here right?

glekner commented:
- Updated the disk summary component
- Added Type column to the VM Disks view

<img width="291" alt="Screen Shot 2019-09-22 at 11 37 27" src="https://user-images.githubusercontent.com/24938324/65384797-40179700-dd2f-11e9-92a0-c4f9bdd8bc55.png">

-------------------

<img width="983" alt="Screen Shot 2019-09-22 at 11 34 32" src="https://user-images.githubusercontent.com/24938324/65384799-4c9bef80-dd2f-11e9-86c1-81c18c89074b.png">

@matthewcarleton - Thoughts?
Also I noticed you color the list items with grey (90,90,90) but in the console all the list items are black like the titles.

matthewcarleton commented:
@glekner I would just leave what the console does for colours. 
Maybe this should be 
```
Drive 1: Container 
kubevirt/cirros-registry-disk-demo
```

glekner commented:
> @glekner I would just leave what the console does for colours.
> Maybe this should be
> 
> ```
> Drive 1: Container 
> kubevirt/cirros-registry-disk-demo
> ```
> 
> This aligns a bit better with what we have in the edit flow. If we go with this we could also just have "Drive 1" instead of "CD-ROM drive 1" in the edit modal. I suppose it's pretty obvious that it's a CD-ROM so we don't need to repeat it. WDYT?

Well the Modal has "CD-ROM" in its title, so I guess putting "Drive" is enough. updating

glekner commented:
changed base branch to 4.3 and rebased

mareklibra commented:
I can see SCSS and JS code here to be copied from #2477 .
Can you please keep the changes in just a single single in either of the PRs and rebase on top of it?

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/_disk-summary.scss:
```diff
@@ -0,0 +1,10 @@
+.kubevirt-disk-summary-disk {
```
Shared with #2477

glekner commented:
closing this PR, we now have one PR for all user stories #2477
